Adds preload scanner post. (#7922)
GoogleChrome/web.dev: src/site/content/en/blog/preload-scanner/index.md

---
title: Don't fight the browser preload scanner
subhead: |
  Find out what the browser preload scanner is, how it helps performance, and how you can stay out of its way.
authors:
  - jlwagner
date: 2022-05-13
hero: image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/ZYfhboCPqPWawQ5LzWKO.jpg
thumbnail: image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/irM6kTXZoP0MmXGtHOSa.jpg
alt: A close-up photograph of a screen with various bits of HTML markup shown in a text editor.
description: |
  Find out what the browser preload scanner is, how it helps performance, and how you can stay out of its way.
tags:
  - blog
  - performance
  - web-vitals
---

> One overlooked aspect of optimizing page speed involves knowing a bit about browser internals. Browsers make certain optimizations for us to improve performance in ways that we as developers can't—but only so long as we don't thwart those optimizations unintentionally.

Reviewer: ```suggestion
One overlooked aspect of optimizing page speed involves knowing a bit about browser internals. Browsers make certain optimizations to improve performance in ways that we as developers can't—but only so long as we don't thwart those optimizations unintentionally.
```

<figure>
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/mXRoJneD6CbMAqaTqNZW.svg", alt="A diagram of how the browser's primary HTML parser parses and tokenizes markup into the DOM. The parser is blocked from doing any further work when it encounters a &lt;link&gt; element pointing to a stylesheet, and can only resume its work until the stylesheet is downloaded and processed.", width="111", height="150" %}
  <figcaption>
    <strong>Fig. 1:</strong> A diagram of how the browser's primary HTML parser can be blocked. In this case, the parser runs into a <code>&lt;link&gt;</code> element for an external CSS file, which blocks the browser from parsing the rest of the document—or even rendering any of it—until the CSS is downloaded and parsed.
  </figcaption>

> </figure>

Reviewer: Note on a11y: A screenreader would read the alt text followed by the caption, thus repeating the info. You have a great caption for this image, so let it do the heavy lifting. I'll make a suggestion for lighter alt text.

Reviewer: ```suggestion
<figure>
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/mXRoJneD6CbMAqaTqNZW.svg", alt="HTML parser diagram.", width="111", height="150" %}
  <figcaption>
    <strong>Fig. 1:</strong> A diagram of how the browser's primary HTML parser can be blocked. In this case, the parser runs into a <code>&lt;link&gt;</code> element for an external CSS file, which blocks the browser from parsing the rest of the document—or even rendering any of it—until the CSS is downloaded and parsed.
  </figcaption>
</figure>
```

In the case of CSS files, both parsing and rendering are blocked in order to prevent a [flash of unstyled content (FOUC)](https://en.wikipedia.org/wiki/Flash_of_unstyled_content), which is when an unstyled version of a page can be seen briefly before styles are applied to it.

> <figure>
>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/8LkMd46pUlwgfYvmTBrO.png", alt="A comparison of the web.dev home page in an unstyled state (left) and the styled state (right). The unstyled state can occur in a flash if the browser doesn't block rendering while a stylesheet is being downloaded and processed.", width="800", height="748", loading="lazy" %}
>   <figcaption>
>     <strong>Fig. 2:</strong> A simulated example of FOUC. At left is the front page of web.dev without styles. At right is the same page with styles applied.
>   </figcaption>
> </figure>

Reviewer: ```suggestion
<figure>
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/8LkMd46pUlwgfYvmTBrO.png", alt="The web.dev home page in an unstyled state (left) and the styled state (right).", width="800", height="748", loading="lazy" %}
  <figcaption>
    <strong>Fig. 2:</strong> A simulated example of FOUC. At left is the front page of web.dev without styles. At right is the same page with styles applied. The unstyled state can occur in a flash if the browser doesn't block rendering while a stylesheet is being downloaded and processed.
  </figcaption>
</figure>
```

Reviewer: That flash detail is interesting and should be seen by all in the caption!

Also, another a11y related alt text note. Some screenreaders have a character limit for alt text of 150, so the shorter the better.

The browser also blocks parsing and rendering of the page when it encounters `<script>` elements without a `defer` or `async` attribute.

{% Aside %}
  Scripts loaded from `<script>` elements with a [`type=module` attribute](https://developer.mozilla.org/docs/Web/JavaScript/Guide/Modules#applying_the_module_to_your_html) are deferred by default.
{% endAside %}

The reason for this is that the browser can't know for sure if any given script will modify the DOM while the primary HTML parser is still doing its job. This is why it's been a common practice to load your JavaScript at the end of the document so that the effects of blocked parsing and rendering become marginal.

These are good reasons for why the browser _should_ block both parsing and rendering, but blocking either of these important steps are desirable, as they can hold up the show by delaying the discovery of other important resources. Thankfully, browsers do their best to mitigate this problem by way of a secondary HTML parser called a _preload scanner_.

<figure>

>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/6lccoVh4f6IJXA8UBKxH.svg", alt="A diagram of both the primary HTML parser (at left) and the preload scanner (at right), which is the secondary HTML parser. At left, the primary HTML parser is blocked processing a stylesheet before it can get to rendering an image, whereas the preload scanner can discover the image resource and begin fetching it while the primary HTML parser is blocked.", width="219", height="150", loading="lazy" %}

Reviewer: ```suggestion
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/6lccoVh4f6IJXA8UBKxH.svg", alt="A diagram of both the primary HTML parser (left) and the preload scanner (right), which is the secondary HTML parser.", width="219", height="150", loading="lazy" %}
```

Reviewer: Same a11y feedback as before, shortened the copy.

  </figcaption>
</figure>

[A preload scanner's role is speculative](https://html.spec.whatwg.org/multipage/parsing.html#speculative-html-parsing), meaning that it examines raw markup in order to find resources to opportunistically fetch before the primary HTML parser would otherwise discover them.

## How to tell when the preload scanner is working

The preload scanner exists _because_ of blocked rendering and parsing. If these two performance issues never existed, the preload scanner wouldn't be very useful. The key to figuring out whether a web page benefits from the preload scanner depends on these blocking phenomena, and to do that, we can introduce an artificial delay for requests to find out where the preload scanner is working.

Let's take [a page](https://preload-scanner-fights.glitch.me/artifically-delayed-requests.html) of basic text and images with a stylesheet. Because CSS files block both rendering and parsing, we can introduce an artificial delay of two seconds for the stylesheet through a proxy service. This delay makes it easier to see in the network waterfall where the preload scanner is working.

<figure>

>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/Gtw08XaoFETKEauBBbBl.png", alt="A WebPageTest network waterfall chart illustrating the effect of the preload scanner. The stylesheet has an artificial delay of two seconds imposed on it before it can load. Even though the delay is large, the preload scanner still finds the image after the stylesheet and begins loading it while the stylesheet blocks the primary HTML parser.", width="800", height="219", loading="lazy" %}

Reviewer: ```suggestion
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/Gtw08XaoFETKEauBBbBl.png", alt="The WebPageTest network waterfall chart illustrates an artificial delay of 2 seconds imposed on the styleesheet.", width="800", height="219", loading="lazy" %}
```

With that toy example out of the way, let's take a look at some real-world patterns where the preload scanner can be defeated—and what can be done to fix them.

{% Aside %}
  This smattering of patterns isn't an exhaustive list, just some common ones.
{% endAside %}

## Injected `async` scripts

Let's say you've got HTML in your `<head>` that includes some inline JavaScript like this:

```html
<script>
  const scriptEl = document.createElement('script');
  scriptEl.src = '/yall.min.js';

  document.head.appendChild(scriptEl);
</script>
```

Injected scripts are [`async`](https://developer.mozilla.org/docs/Web/HTML/Element/script#attr-async) by default, so when this script is injected, it will behave as if the `async` attribute was applied to it. That means it will run as soon as possible and not block rendering. Sounds optimal, right? Yet, if we presume that this inline `<script>` comes after a `<link>` element that loads an external CSS file, we get a suboptimal result:

<figure>

>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/L9ltNSHa6D4FI6YPw1vF.png", alt="A WebPageTest network waterfall chart showing how the preload scanner is defeated in discovering a script that is injected with inline JavaScript. The browser can't detect the script while the stylesheet blocks the primary HTML parser, and delays script discovery by about 600 milliseconds.", width="800", height="219", loading="lazy" %}

Reviewer: ```suggestion
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/L9ltNSHa6D4FI6YPw1vF.png", alt="This WebPageTest chart shows the preload scan defeated when a script is injected.", width="800", height="219", loading="lazy" %}
```
